util/latlng2place_mapsapi.py

```diff
@@ -0,0 +1,171 @@
+  ... more calls to llp.resolve() ...
+"""
+
+import os
+import requests
+import sys
+import threading
+
+# Allows the following module imports to work when running as a script
+sys.path.append(
+    os.path.dirname(os.path.dirname(os.path.dirname(
+        os.path.abspath(__file__)))))
+
+import aa_isocode2dcid
+import alpha2_to_dcid
+
+_LOCK = threading.Lock()
+_PLACE_TYPES = ['administrative_area2', 'administrative_area1', 'country']
+_RECON_ROOT = 'https://api.datacommons.org/v1/recon/resolve/id'
+
+
+def _call_rpc(url):
+    return requests.get(url, timeout=30).json()
+
+
+def _placeid2dcid(place_id):
+    resp = requests.post(_RECON_ROOT,
+                         json={
+                             'in_prop': 'placeId',
+                             'out_prop': 'dcid',
+                             'ids': [place_id]
+                         })
+    resp.raise_for_status()
+    resp_json = resp.json()
+    if 'entities' in resp_json and resp_json['entities']:
+        return resp_json['entities'][0]['outIds'][0]
+    return ''
+
+
+class Resolver(object):
+    """Resolves lat lngs to DCIDs."""
+
+    def __init__(self, api_key, cache_file='', cache_only=False):
+        assert api_key, 'Resolver() needs a valid API key'
+        self._api_key = api_key
+        self._cache_file = cache_file
+        self._cache_only = cache_only
+        self._latlng2place = {}
+        if cache_file:
+            try:
+                with open(cache_file, 'r') as cf:
+                    for line in cf:
+                        parts = line.strip().split(',')
+                        if len(parts) < 3:
+                            continue
+                        latlng = parts[0] + ',' + parts[1]
+                        self._latlng2place[latlng] = parts[2:]
+            except Exception:
+                pass
+            self._cf = open(cache_file, 'a+')
+
+    def _url(self):
+        return 'https://maps.googleapis.com/maps/api/geocode/json?sensor=false&key=' + self._api_key
+
+    def _parse_result(self, result):
+        """Parse resultsf rom Cloud reverse geocoding service."""
+        # Find the placeId in the result.
+        if 'place_id' not in result or 'types' not in result:
+            return []
+
+        place_type = ''
+        for pt in _PLACE_TYPES:
+            mt = pt.replace('_area', '_area_level_')
+            if mt in result['types']:
+                place_type = pt
+                break
+        place_id = result['place_id']
+
+        # Extract short-names.
+        short_names = {}
+        for place in result['address_components']:
+            if 'administrative_area_level_2' in place['types']:
+                short_names['administrative_area2'] = place['short_name']
+            elif 'administrative_area_level_1' in place['types']:
+                short_names['administrative_area1'] = place['short_name']
+            elif 'country' in place['types']:
+                short_names['country'] = place['short_name']
```

Comment:
i wonder if we could do something like:

```
for place in result['address_components']:
  for pt in _PLACE_TYPES:
    if pt in place['types']
      short_names[pt] = place['short_name']
      break
```
noting the ordering constraint in `_PLACE_TYPES`

Reply:
Good idea!  (I'd largely blindly ported http://google3/datacommons/mcf/lib/latlng2place.py;l=64-103;rcl=481245757)